File: scripts/bundle_runtime.js
'use strict';

const fs   = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const RESOURCES = path.join(ROOT, 'resources');

// ── Download helper (follows redirects) ─────────────────────────────────────
function _downloadOne(url, destPath) {
  return new Promise((resolve, reject) => {
    const proto = url.startsWith('https') ? require('https') : require('http');
    const req = proto.get(url, { headers: { 'User-Agent': 'Vellum/1.0' } }, res => {
      if ([301, 302, 303, 307, 308].includes(res.statusCode) && res.headers.location) {
        let next = res.headers.location;
        if (!next.startsWith('http')) {
          const u = require('url').parse(url);
          next = `${u.protocol}//${u.host}${next}`;
        }
        res.resume();
        resolve(_downloadOne(next, destPath)); // follow redirect
        return;
      }
      if (res.statusCode !== 200) {
        reject(new Error(`HTTP ${res.statusCode}`));
        return;
      }
      const total = parseInt(res.headers['content-length'], 10) || 0;
      let loaded = 0;
      const file = fs.createWriteStream(destPath);
      res.on('data', chunk => {
        loaded += chunk.length;
        if (total > 0 && loaded % (5 * 1024 * 1024) < 65536) {
          process.stdout.write(`\r    ${Math.round(loaded/total*100)}%  ${(loaded/1e6).toFixed(0)}/${(total/1e6).toFixed(0)} MB  `);
        }
      });
      res.pipe(file);
      file.on('finish', () => { file.close(); if (total > 0) process.stdout.write('\n'); resolve(); });
      file.on('error', reject);
    });
    req.on('error', reject);
    req.setTimeout(60000, () => { req.destroy(); reject(new Error('timeout')); });
  });
}

async function download(url, destPath, label) {
  if (fs.existsSync(destPath) && fs.statSync(destPath).size > 1000) {
    console.log(`  [skip] ${label} — already present`);
    return;
  }
  console.log(`  [downloading] ${label}…`);
  try {
    await _downloadOne(url, destPath);
    const size = fs.statSync(destPath).size;
    if (size < 1000) throw new Error('file too small');
    console.log(`  [done] ${label} (${(size/1e6).toFixed(1)} MB)`);
  } catch (e) {
    try { fs.unlinkSync(destPath); } catch (_) {}
    throw new Error(`Download failed for ${label}: ${e.message}\nURL: ${url}`);
  }
}

// ── Zip extraction ──────────────────────────────────────────────────────────
async function extractZip(zipPath, destDir) {
  // Use the built-in approach — spawn powershell Expand-Archive on Windows,
  // or try a pure-Node approach if we have an unzip utility.
  // Simplest cross-platform: use Node's built-in zlib + a simple zip reader.
  const { spawnSync } = require('child_process');
  if (process.platform === 'win32') {
    // PowerShell Expand-Archive is always available on modern Windows
    const r = spawnSync('powershell', [
      '-NoProfile', '-Command',
      `Expand-Archive -LiteralPath '${zipPath}' -DestinationPath '${destDir}' -Force`
    ], { stdio: 'pipe' });
    if (r.status !== 0) throw new Error('Failed to extract zip: ' + r.stderr.toString());
  } else {
    // macOS / Linux: use unzip
    const r = spawnSync('unzip', ['-o', '-q', zipPath, '-d', destDir], { stdio: 'pipe' });
    if (r.status !== 0) throw new Error('Failed to extract zip: ' + r.stderr.toString());
  }
}

// ── tar.gz extraction ───────────────────────────────────────────────────────
async function extractTarGz(tarPath, destDir) {
  const { spawnSync } = require('child_process');
  const r = spawnSync('tar', ['-xzf', tarPath, '-C', destDir], { stdio: 'pipe' });
  if (r.status !== 0) throw new Error('Failed to extract tar.gz: ' + r.stderr.toString());
}

// ── Windows: embedded Python + uv ───────────────────────────────────────────
const PYTHON_VERSION = '3.11.9';
const UV_VERSION     = '0.11.14';

async function bundleWindows() {
  console.log('\n── Bundling Windows runtime ──');

  const pyDir  = path.join(RESOURCES, 'python');
  const toolDir = path.join(RESOURCES, 'tools');

  fs.mkdirSync(pyDir,   { recursive: true });
  fs.mkdirSync(toolDir, { recursive: true });

  // Embedded Python (no installer needed — just extract and use)
  const pyZip = path.join(RESOURCES, 'python-embed.zip');
  const pyUrl = `https://www.python.org/ftp/python/${PYTHON_VERSION}/python-${PYTHON_VERSION}-embed-amd64.zip`;

  await download(pyUrl, pyZip, 'Python embeddable');
  if (!fs.existsSync(path.join(pyDir, 'python.exe'))) {
    console.log('  [extracting] Python…');
    await extractZip(pyZip, pyDir);
    console.log('  [done] Python extracted');
  }

  // Configure embedded Python to find packages in the standard location
  // The embeddable Python uses a ._pth file to control import paths
  const [pyMajor, pyMinor] = PYTHON_VERSION.split('.');
  const pthFile = path.join(pyDir, `python${pyMajor}${pyMinor}._pth`);
  const pthContent = `python${pyMajor}${pyMinor}.zip
.
Lib
Lib\\site-packages
import site
`;
  fs.writeFileSync(pthFile, pthContent, 'utf8');
  console.log(`  [done] Python ._pth configured (${path.basename(pthFile)})`);

  // Ensure Lib/site-packages exists
  fs.mkdirSync(path.join(pyDir, 'Lib', 'site-packages'), { recursive: true });

  // Bootstrap pip into the embedded Python (ensurepip is stripped from embeddable)
  const pipZip = path.join(pyDir, 'get-pip.py');
  await download('https://bootstrap.pypa.io/get-pip.py', pipZip, 'get-pip.py');

  console.log('  [bootstrapping] pip into embedded Python…');
  const { execFileSync } = require('child_process');
  try {
    execFileSync(path.join(pyDir, 'python.exe'), [pipZip, '--no-warn-script-location'], {
      cwd: pyDir,
      timeout: 60000,
      stdio: 'pipe',
    });
    console.log('  [done] pip installed');
  } catch (e) {
    console.error('  [warn] pip bootstrap failed:', e.stderr ? e.stderr.toString() : e.message);
    console.error('  [warn] setup will bootstrap pip on first launch');
  }

  // Cleanup get-pip.py — keep it for runtime fallback
  // (it gets excluded from the installer via !get-pip.py filter)

  // uv binary for Windows
  const uvZip = path.join(RESOURCES, 'uv-win.zip');
  const uvUrl = `https://github.com/astral-sh/uv/releases/download/${UV_VERSION}/uv-x86_64-pc-windows-msvc.zip`;
  await download(uvUrl, uvZip, 'uv (Windows)');

  if (!fs.existsSync(path.join(toolDir, 'uv.exe'))) {
    console.log('  [extracting] uv…');
    await extractZip(uvZip, toolDir);
    console.log('  [done] uv extracted');
  }

  // Cleanup
  try { fs.unlinkSync(pyZip); } catch (_) {}
  try { fs.unlinkSync(uvZip); } catch (_) {}
  try { fs.unlinkSync(pipZip); } catch (_) {}
}

// ── macOS: python-build-standalone + uv ─────────────────────────────────────
const STANDALONE_VERSION = '20260602';

async function bundleMac(arch) {
  const label = arch === 'arm64' ? 'Apple Silicon' : 'Intel';
  console.log(`\n── Bundling macOS runtime (${label}) ──`);

  const pyDir  = path.join(RESOURCES, 'python');   // same dir as Windows — extraResources expects this
  const toolDir = path.join(RESOURCES, 'tools');

  fs.mkdirSync(pyDir,   { recursive: true });
  fs.mkdirSync(toolDir, { recursive: true });

  // python-build-standalone
  const pyArch = arch === 'arm64' ? 'aarch64' : 'x86_64';
  const standaloneName = `cpython-3.11.15+${STANDALONE_VERSION}-${pyArch}-apple-darwin-install_only`;
  const pyTar = path.join(RESOURCES, 'python-mac.tar.gz');
  const pyUrl = `https://github.com/astral-sh/python-build-standalone/releases/download/${STANDALONE_VERSION}/${standaloneName}.tar.gz`;

  await download(pyUrl, pyTar, `Python standalone (macOS ${label})`);

  if (!fs.existsSync(path.join(pyDir, 'bin', 'python3'))) {
    console.log('  [extracting] Python…');
    await extractTarGz(pyTar, pyDir);
    // python-build-standalone tars wrap everything in a 'python/' directory.
    // Move files up one level if needed.
    const inner = path.join(pyDir, 'python');
    if (fs.existsSync(inner) && fs.statSync(inner).isDirectory()) {
      for (const f of fs.readdirSync(inner)) {
        const src = path.join(inner, f);
        const dst = path.join(pyDir, f);
        try { fs.renameSync(src, dst); } catch (_) { /* cross-device? copy+delete */ }
      }
      try { fs.rmdirSync(inner); } catch (_) {}
    }
    if (!fs.existsSync(path.join(pyDir, 'bin', 'python3'))) {
      throw new Error(`Python extraction failed: bin/python3 not found in ${pyDir}`);
    }
    console.log('  [done] Python extracted');
  }

  // Install pip into standalone Python
  const pipPy = path.join(pyDir, 'get-pip.py');
  await download('https://bootstrap.pypa.io/get-pip.py', pipPy, 'get-pip.py');

  // uv binary for macOS
  const uvArch = arch === 'arm64' ? 'aarch64' : 'x86_64';
  const uvTar = path.join(RESOURCES, 'uv-mac.tar.gz');
  const uvUrl = `https://github.com/astral-sh/uv/releases/download/${UV_VERSION}/uv-${uvArch}-apple-darwin.tar.gz`;
  await download(uvUrl, uvTar, `uv (macOS ${label})`);

  if (!fs.existsSync(path.join(toolDir, 'uv'))) {
    console.log('  [extracting] uv…');
    await extractTarGz(uvTar, toolDir);
    // uv tars may have a uv-{arch}-apple-darwin/ wrapper directory
    const uvDirs = fs.readdirSync(toolDir).filter(f => f.startsWith('uv-') && fs.statSync(path.join(toolDir, f)).isDirectory());
    for (const d of uvDirs) {
      const srcDir = path.join(toolDir, d);
      for (const f of fs.readdirSync(srcDir)) {
        const dst = path.join(toolDir, f);
        try { fs.renameSync(path.join(srcDir, f), dst); } catch (_) {}
      }
      try { fs.rmdirSync(srcDir); } catch (_) {}
    }
    // Make executable
    try { fs.chmodSync(path.join(toolDir, 'uv'), 0o755); } catch (_) {}
    if (!fs.existsSync(path.join(toolDir, 'uv'))) {
      throw new Error('uv extraction failed: uv binary not found');
    }
    console.log('  [done] uv extracted');
  }

  // Cleanup
  try { fs.unlinkSync(pyTar); } catch (_) {}
  try { fs.unlinkSync(uvTar); } catch (_) {}
  try { fs.unlinkSync(pipPy); } catch (_) {}
}

// ── Main ────────────────────────────────────────────────────────────────────
async function main() {
  const args = process.argv.slice(2);
  const doAll = args.includes('--all');
  const doWin = doAll || args.includes('--win') || (!doAll && process.platform === 'win32' && args.length === 0);
  const doMac = doAll || args.includes('--mac') || (!doAll && process.platform === 'darwin' && args.length === 0);

  if (args.length === 0 && process.platform !== 'win32' && process.platform !== 'darwin') {
    console.log('Unsupported platform. Use --win or --mac.');
    process.exit(1);
  }

  if (doWin) await bundleWindows();
  if (doMac) {
    await bundleMac('arm64');
    await bundleMac('x64');
  }

  console.log('\n✓ Runtime bundle complete\n');
}

main().catch(err => {
  console.error('\n✗ Bundle failed:', err.message);
  process.exit(1);
});

File: scripts/bundle_runtime.py
#!/usr/bin/env python3
"""Bundle runtime + build dependencies for Vellum.
Downloads and extracts portable Python (for the end-user app), uv, and Node.js
(for the build step only).

Usage:
  python scripts/bundle_runtime.py --win      # Windows runtime + Node
  python scripts/bundle_runtime.py --mac      # macOS runtime + Node
  python scripts/bundle_runtime.py --all      # all platforms
  python scripts/bundle_runtime.py --node     # Node.js only (if runtime already bundled)
"""

import argparse
import os
import shutil
import subprocess
import sys
import tarfile
import urllib.request
import zipfile
from pathlib import Path
from typing import Optional

ROOT = Path(__file__).resolve().parent.parent
RESOURCES = ROOT / "resources"

PYTHON_EMBED_VERSION = "3.11.9"
STANDALONE_VERSION = "20260602"
UV_VERSION = "0.11.14"
NODE_VERSION = "22.14.0"

CHUNK_SIZE = 8192


def _download(url: str, dest: Path, label: str = "") -> None:
    if dest.exists() and dest.stat().st_size > 1000:
        print(f"  [skip] {label or dest.name}")
        return
    print(f"  [download] {label or dest.name}")
    tmp = dest.with_suffix(dest.suffix + ".tmp")
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "Vellum/1.0"})
        with urllib.request.urlopen(req, timeout=120) as resp:
            total = int(resp.headers.get("Content-Length", 0))
            loaded = 0
            with open(tmp, "wb") as f:
                while True:
                    chunk = resp.read(CHUNK_SIZE)
                    if not chunk:
                        break
                    f.write(chunk)
                    loaded += len(chunk)
                    if total > 0 and loaded % (5 * 1024 * 1024) < 65536:
                        pct = loaded * 100 // total
                        sys.stdout.write(f"\r    {pct}%  {loaded // 1048576} / {total // 1048576} MB  ")
                        sys.stdout.flush()
            if total > 0:
                print()
        tmp.rename(dest)
        mb = dest.stat().st_size / 1048576
        print(f"  [done] {mb:.0f} MB")
    except Exception as e:
        tmp.unlink(missing_ok=True)
        raise RuntimeError(f"Download failed for {label or url}") from e


def _extract_zip(zip_path: Path, dest: Path) -> None:
    print(f"  [extract] {zip_path.name}")
    with zipfile.ZipFile(zip_path, "r") as zf:
        zf.extractall(dest)


def _extract_targz(tar_path: Path, dest: Path) -> None:
    print(f"  [extract] {tar_path.name}")
    with tarfile.open(tar_path, "r:gz") as tf:
        tf.extractall(dest)


def _flatten_wrapper_dir(dest: Path, marker: str) -> None:
    """Move files out of a single wrapper directory (e.g. `python/` inside `python/`)."""
    for child in list(dest.iterdir()):
        if child.is_dir() and child.name.startswith(marker):
            for f in child.iterdir():
                dst = dest / f.name
                if dst.exists():
                    (shutil.rmtree if dst.is_dir() else dst.unlink)(dst)
                shutil.move(str(f), str(dst))
            shutil.rmtree(child)


def _bootstrap_pip(python_bin: Path, pip_py: Path) -> None:
    print("  [bootstrap] pip")
    result = subprocess.run(
        [str(python_bin), str(pip_py), "--no-warn-script-location"],
        cwd=str(python_bin.parent),
        timeout=60,
        capture_output=True,
        text=True,
    )
    if result.returncode == 0:
        print("  [done] pip installed")
    else:
        print(f"  [warn] pip bootstrap failed: {result.stderr.strip()[-200:]}")


# ── Windows ──────────────────────────────────────────────────────────────────

def bundle_windows() -> None:
    print("\n── Bundling Windows runtime ──")
    py_dir = RESOURCES / "python"
    tool_dir = RESOURCES / "tools"
    py_dir.mkdir(parents=True, exist_ok=True)
    tool_dir.mkdir(parents=True, exist_ok=True)

    py_zip = RESOURCES / "python-embed.zip"
    _download(
        f"https://www.python.org/ftp/python/{PYTHON_EMBED_VERSION}/python-{PYTHON_EMBED_VERSION}-embed-amd64.zip",
        py_zip, "Python embeddable",
    )
    if not (py_dir / "python.exe").exists():
        _extract_zip(py_zip, py_dir)

    maj, minor = PYTHON_EMBED_VERSION.split(".")[0], PYTHON_EMBED_VERSION.split(".")[1]
    pth = py_dir / f"python{maj}{minor}._pth"
    pth.write_text(
        f"python{maj}{minor}.zip\n.\nLib\nLib\\site-packages\nimport site\n"
    )
    (py_dir / "Lib" / "site-packages").mkdir(parents=True, exist_ok=True)
    print(f"  [done] {pth.name} configured")

    pip_py = py_dir / "get-pip.py"
    _download("https://bootstrap.pypa.io/get-pip.py", pip_py, "get-pip.py")
    _bootstrap_pip(py_dir / "python.exe", pip_py)

    uv_zip = RESOURCES / "uv-win.zip"
    _download(
        f"https://github.com/astral-sh/uv/releases/download/{UV_VERSION}/uv-x86_64-pc-windows-msvc.zip",
        uv_zip, "uv (Windows)",
    )
    if not (tool_dir / "uv.exe").exists():
        _extract_zip(uv_zip, tool_dir)
        _flatten_wrapper_dir(tool_dir, "uv-")

    for p in [py_zip, uv_zip, pip_py]:
        p.unlink(missing_ok=True)


# ── macOS ────────────────────────────────────────────────────────────────────

def bundle_mac(arch: str) -> None:
    label = "Apple Silicon" if arch == "arm64" else "Intel"
    py_arch = "aarch64" if arch == "arm64" else "x86_64"
    print(f"\n── Bundling macOS runtime ({label}) ──")

    py_dir = RESOURCES / "python"
    tool_dir = RESOURCES / "tools"
    py_dir.mkdir(parents=True, exist_ok=True)
    tool_dir.mkdir(parents=True, exist_ok=True)

    standalone = f"cpython-3.11.15+{STANDALONE_VERSION}-{py_arch}-apple-darwin-install_only"
    py_tar = RESOURCES / "python-mac.tar.gz"
    _download(
        f"https://github.com/astral-sh/python-build-standalone/releases/download/{STANDALONE_VERSION}/{standalone}.tar.gz",
        py_tar, f"Python standalone (macOS {label})",
    )
    if not (py_dir / "bin" / "python3").exists():
        _extract_targz(py_tar, py_dir)
        _flatten_wrapper_dir(py_dir, "python")
        if not (py_dir / "bin" / "python3").exists():
            raise RuntimeError("Python extraction failed: bin/python3 not found")
        print("  [done] Python extracted")

    pip_py = py_dir / "get-pip.py"
    _download("https://bootstrap.pypa.io/get-pip.py", pip_py, "get-pip.py")
    _bootstrap_pip(py_dir / "bin" / "python3", pip_py)

    uv_arch = "aarch64" if arch == "arm64" else "x86_64"
    uv_tar = RESOURCES / "uv-mac.tar.gz"
    _download(
        f"https://github.com/astral-sh/uv/releases/download/{UV_VERSION}/uv-{uv_arch}-apple-darwin.tar.gz",
        uv_tar, f"uv (macOS {label})",
    )
    if not (tool_dir / "uv").exists():
        _extract_targz(uv_tar, tool_dir)
        _flatten_wrapper_dir(tool_dir, "uv-")
        (tool_dir / "uv").chmod(0o755)
        if not (tool_dir / "uv").exists():
            raise RuntimeError("uv extraction failed: uv binary not found")
        print("  [done] uv extracted")

    for p in [py_tar, uv_tar, pip_py]:
        p.unlink(missing_ok=True)


# ── Node.js (for build step) ─────────────────────────────────────────────────

def bundle_node(platform: str, arch: Optional[str] = None) -> None:
    print(f"\n── Bundling Node.js v{NODE_VERSION} ({platform}) ──")
    node_dir = RESOURCES / "node"
    node_dir.mkdir(parents=True, exist_ok=True)

    if platform == "win":
        archive = RESOURCES / "node-win.zip"
        url = f"https://nodejs.org/dist/v{NODE_VERSION}/node-v{NODE_VERSION}-win-x64.zip"
        _download(url, archive, "Node.js (Windows)")
        if not (node_dir / "node.exe").exists():
            _extract_zip(archive, node_dir)
            _flatten_wrapper_dir(node_dir, "node-v")
            (node_dir / "npm.cmd").chmod(0o755)
            (node_dir / "npx.cmd").chmod(0o755)
            print("  [done] Node.js ready")
    else:
        arch_suffix = "arm64" if arch == "arm64" else "x64"
        archive = RESOURCES / f"node-mac-{arch_suffix}.tar.gz"
        url = f"https://nodejs.org/dist/v{NODE_VERSION}/node-v{NODE_VERSION}-darwin-{arch_suffix}.tar.gz"
        _download(url, archive, f"Node.js (macOS {arch_suffix})")
        if not (node_dir / "bin" / "node").exists():
            _extract_targz(archive, node_dir)
            _flatten_wrapper_dir(node_dir, "node-v")
            (node_dir / "bin" / "node").chmod(0o755)
            (node_dir / "bin" / "npm").chmod(0o755)
            if not (node_dir / "bin" / "node").exists():
                raise RuntimeError("Node.js extraction failed")
            print("  [done] Node.js ready")

    archive.unlink(missing_ok=True)


# ── Main ─────────────────────────────────────────────────────────────────────

def main() -> None:
    parser = argparse.ArgumentParser(description="Bundle Vellum build/runtime dependencies")
    parser.add_argument("--win", action="store_true", help="Bundle Windows runtime + Node.js")
    parser.add_argument("--mac", action="store_true", help="Bundle macOS runtime + Node.js")
    parser.add_argument("--all", action="store_true", help="Bundle all platforms")
    parser.add_argument("--node", action="store_true", help="Bundle Node.js only")
    args = parser.parse_args()

    do_win = args.win or args.all
    do_mac = args.mac or args.all

    if args.node:
        bundle_node("win")
        bundle_node("mac", "arm64")
        bundle_node("mac", "x64")
        print("\n✓ Node.js bundle complete\n")
        return

    if not (do_win or do_mac):
        if sys.platform == "win32":
            do_win = True
        elif sys.platform == "darwin":
            do_mac = True
        else:
            parser.print_help()
            sys.exit(1)

    if do_win:
        bundle_windows()
        bundle_node("win")
    if do_mac:
        bundle_mac("arm64")
        bundle_mac("x64")
        # Node.js for the build machine's native arch only
        native_arch = "arm64" if os.uname().machine in ("arm64", "aarch64") else "x64" if sys.platform == "darwin" else "x64"
        bundle_node("mac", native_arch)

    print("\n✓ Bundle complete\n")


if __name__ == "__main__":
    main()
